### PR TITLE
Promote fusing select-like ops with producers

### PR DESCRIPTION
**Pull request by naoyam**
For #278, our schedulers currently can't fuse them into a single kernel, so a segmentation needs to happen. Currently, it's segmented between softmax and take_along_axis, just because of the ordering of the segmenter. However, we want the take_along_axis op to be fused together with the preceding softmax since then the temporary output from the first segment would be much smaller, reducing gmem access overhead. In the case of `[8192, 32768]`, the (logical) I/O cost would be `1 / 32768`.

This PR introduces a simple mechanism to allow preferred fusions in the segmentation steps. Currently, there's only preference of select-like ops with producers. "Select-like" here also includes index_select and torch_gather to size-one domains. In those ops, it's guaranteed that the size of the consumer tensor is no larger than the lookup tensor, so it makes sense to fuse them with producers.

Currently, it's only tested with the cross-entropy loss case. The overall segmentation algorithm would need to go through significant refactoring, so I don't think making this interface super robust is worth doing at this moment, and it's likely to be redesigned. For now, this is very important for the cross-entropy performance.

### CODE REVIEW

File: csrc/fusion_segmenter.cpp [hunk 1]
```diff
@@ -3123,34 +3261,14 @@ void SegmentCandidateFinder::findSegments() {
...
-        group->merge_through_ = candidate_it->edge;
+      // Try preferred merge first
+      for (auto& [group, neighbor] :
+           PreferredMergeCandidatePicker::get(groups())) {
+        trySetUpMerge(group, {neighbor});
+      }
 
-        candidate_it->group->merged_ = true;
-        candidate_it->group->merge_with_ = group;
-        candidate_it->group->merge_through_ = candidate_it->edge;
```
Comment (naoyam) [inline csrc/fusion_segmenter.cpp:3153]: This part is refactored into `trySetupMerge`

File: csrc/fusion_segmenter.cpp [hunk 2]
```diff
@@ -2973,6 +2973,108 @@ bool areDirectlyConnected(SegmentedGroup* group1, SegmentedGroup* group2) {
...
+  }
+
+  // Find the producer group that corresponds to the lookup tensor
+  // of the expr.
+  auto producer_edge_it = std::find_if(
+      group->producer_edges.begin(),
+      group->producer_edges.end(),
+      [&lookup_tv](SegmentedEdge* edge) { return edge->val == lookup_tv; });
+
+  // Not sure this could happen happen. Just assert for now.
```
Comment (zasdfgbnm) [inline csrc/fusion_segmenter.cpp:3068]:
```suggestion
  // Not sure this could happen. Just assert for now.
```

File: csrc/fusion_segmenter.cpp [hunk 3]
```diff
@@ -3123,34 +3261,14 @@ void SegmentCandidateFinder::findSegments() {
...
+      for (auto& [group, neighbor] :
+           PreferredMergeCandidatePicker::get(groups())) {
+        trySetUpMerge(group, {neighbor});
+      }
 
-        candidate_it->group->merged_ = true;
-        candidate_it->group->merge_with_ = group;
-        candidate_it->group->merge_through_ = candidate_it->edge;
+      for (auto& group : groups()) {
+        trySetUpMerge(group);
```
Comment (zasdfgbnm) [inline csrc/fusion_segmenter.cpp:3271]: Should we skip this if `to_merge_` is not empty here, so that we actually apply the preferred merge and reevaluate if there are new preferred merges? I know that this won't make a difference today, but if in the future we want to generalize `PreferredMergeCandidatePicker` for other cases, I think this behavior makes more sense.

**Review by zasdfgbnm — APPROVED** [review]
Left some comments, no real blockers.

File: csrc/fusion_segmenter.cpp [hunk 4]
```diff
@@ -2973,6 +2973,108 @@ bool areDirectlyConnected(SegmentedGroup* group1, SegmentedGroup* group2) {
...
+  std::vector<std::pair<SegmentedGroup*, SegmentedGroup::NeighborGroup>>
+      candidates_;
+};
+
+std::optional<SegmentedGroup::NeighborGroup> PreferredMergeCandidatePicker::
+    mergeSelectLikeOpsWithProducers(SegmentedGroup* group) const {
+  const auto& exprs = group->exprs();
+
+  // I *think* it's enough to consider the initial merge of
+  // select-like ops.
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/fusion_segmenter.cpp:3028]: I expect we would iterate these heuristics several times to make it really robust, but in your case, it would first merge `take_along_axis` and `neg` as a single group, then the segmenter would have four groups: `{take_along_axis, neg}`, `{log}`, `{sum}`, and `{exp}`. (There's additional pre-main-loop heuristics to group reductions, which is ignored here). Then, it would just loop over these groups and merge possible pairs of groups. In this particular case, since `{exp}` is the first group, `{exp}` and `{sum}` would be merged first, and then `{log}` and `{take_along_axis, neg}` would be merged together, resulting in two groups: `{sum, exp}` and `{take_along_axis, neg, log}`. If these final two groups would be merged or not would depend on the reduction scheduler as it may not be able to have the exprs in the another group as post-reduction ops. If allowed, all would be grouped as a signle fusion, or otherwise, the last two groups would be presented as the final segmented fusions.

**Comment by naoyam** [comment]
!build
